File: src/client/java/forfun/miningqol/client/gui/NameHiderCategoryScreen.java
package forfun.miningqol.client.gui;

import forfun.miningqol.client.MiningqolClient;
import forfun.miningqol.client.NameHider;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.gui.widget.TextFieldWidget;
import net.minecraft.text.Text;

import java.util.ArrayList;
import java.util.List;

/**
 * Category screen for Name Hider settings
 */
public class NameHiderCategoryScreen extends Screen {
    private static final int WINDOW_WIDTH = 550;
    private static final int WINDOW_HEIGHT = 450;
    private static final int CONTROL_WIDTH = 400;
    private static final int CONTROL_HEIGHT = 35;
    private static final int CONTROL_SPACING = 8;
    private static final int BACK_BUTTON_WIDTH = 80;
    private static final int BACK_BUTTON_HEIGHT = 30;
    private static final int SLIDER_WIDTH = 120;
    private static final int TEXT_FIELD_HEIGHT = 20;

    private final Screen parent;
    private List<Control> controls;
    private int hoveredControl = -1;
    private boolean backButtonHovered = false;
    private int draggingSlider = -1;
    private TextFieldWidget nameField;

    private int windowX, windowY;

    private interface Control {
        void render(DrawContext context, int x, int y, boolean hovered, int mouseX);
        boolean click(int mouseX, int mouseY, int x, int y);
        void drag(int mouseX, int x);
        int getHeight();
    }

    private static class ToggleControl implements Control {
        final String label;
        final Runnable onToggle;
        boolean enabled;

        ToggleControl(String label, boolean enabled, Runnable onToggle) {
            this.label = label;
            this.enabled = enabled;
            this.onToggle = onToggle;
        }

        @Override
        public void render(DrawContext context, int x, int y, boolean hovered, int mouseX) {
            int bgColor = hovered ? 0xCC303030 : 0xCC242424;
            context.fill(x, y, x + CONTROL_WIDTH, y + CONTROL_HEIGHT, bgColor);
            context.drawBorder(x, y, CONTROL_WIDTH, CONTROL_HEIGHT, hovered ? 0xFF4488FF : 0xFF404040);

            int indicatorColor = enabled ? 0xFF44FF44 : 0xFF404040;
            context.fill(x + 3, y + 3, x + 9, y + CONTROL_HEIGHT - 3, indicatorColor);

            int labelX = x + 15;
            int labelY = y + (CONTROL_HEIGHT - 8) / 2;
            net.minecraft.client.font.TextRenderer textRenderer = net.minecraft.client.MinecraftClient.getInstance().textRenderer;
            context.drawTextWithShadow(textRenderer, label, labelX, labelY, 0xFFFFFF);

            String statusText = enabled ? "ON" : "OFF";
            int statusX = x + CONTROL_WIDTH - textRenderer.getWidth(statusText) - 10;
            context.drawTextWithShadow(textRenderer, statusText, statusX, labelY, enabled ? 0x44FF44 : 0x808080);
        }

        @Override
        public boolean click(int mouseX, int mouseY, int x, int y) {
            if (mouseX >= x && mouseX <= x + CONTROL_WIDTH && mouseY >= y && mouseY <= y + CONTROL_HEIGHT) {
                onToggle.run();
                return true;
            }
            return false;
        }

        @Override
        public void drag(int mouseX, int x) {}

        @Override
        public int getHeight() { return CONTROL_HEIGHT; }
    }

    private static class TextFieldControl implements Control {
        final String label;
        TextFieldWidget textField;

        TextFieldControl(String label, TextFieldWidget textField) {
            this.label = label;
            this.textField = textField;
        }

        @Override
        public void render(DrawContext context, int x, int y, boolean hovered, int mouseX) {
            int bgColor = hovered ? 0xCC303030 : 0xCC242424;
            context.fill(x, y, x + CONTROL_WIDTH, y + CONTROL_HEIGHT, bgColor);
            context.drawBorder(x, y, CONTROL_WIDTH, CONTROL_HEIGHT, hovered ? 0xFF4488FF : 0xFF404040);

            net.minecraft.client.font.TextRenderer textRenderer = net.minecraft.client.MinecraftClient.getInstance().textRenderer;
            int labelX = x + 10;
            int labelY = y + 5;
            context.drawTextWithShadow(textRenderer, label, labelX, labelY, 0xFFFFFF);

            // Position text field
            textField.setX(x + 10);
            textField.setY(y + 17);
            textField.setWidth(CONTROL_WIDTH - 20);
            textField.render(context, mouseX, y, 0);
        }

        @Override
        public boolean click(int mouseX, int mouseY, int x, int y) {
            boolean clicked = textField.mouseClicked(mouseX, mouseY, 0);
            if (clicked) {
                textField.setFocused(true);
            }
            return clicked;
        }

        @Override
        public void drag(int mouseX, int x) {}

        @Override
        public int getHeight() { return CONTROL_HEIGHT; }
    }

    private static class ColorSliderControl implements Control {
        final String label;
        final String channelName;
        final Runnable onChange;
        float value;

        ColorSliderControl(String label, String channelName, float value, Runnable onChange) {
            this.label = label;
            this.channelName = channelName;
            this.value = value;
            this.onChange = onChange;
        }

        @Override
        public void render(DrawContext context, int x, int y, boolean hovered, int mouseX) {
            int bgColor = hovered ? 0xCC303030 : 0xCC242424;
            context.fill(x, y, x + CONTROL_WIDTH, y + CONTROL_HEIGHT, bgColor);
            context.drawBorder(x, y, CONTROL_WIDTH, CONTROL_HEIGHT, hovered ? 0xFF4488FF : 0xFF404040);

            net.minecraft.client.font.TextRenderer textRenderer = net.minecraft.client.MinecraftClient.getInstance().textRenderer;
            int labelX = x + 10;
            int labelY = y + 6;
            context.drawTextWithShadow(textRenderer, label + " - " + channelName, labelX, labelY, 0xFFFFFF);

            // Slider bar
            int sliderX = x + 10;
            int sliderY = y + 20;
            context.fill(sliderX, sliderY, sliderX + SLIDER_WIDTH, sliderY + 8, 0xFF1A1A1A);
            context.drawBorder(sliderX, sliderY, SLIDER_WIDTH, 8, 0xFF404040);

            // Slider fill with color preview
            int fillWidth = (int) (SLIDER_WIDTH * value);
            int previewColor = getPreviewColor();
            context.fill(sliderX + 1, sliderY + 1, sliderX + fillWidth - 1, sliderY + 7, previewColor);

            // Slider handle
            int handleX = sliderX + fillWidth - 4;
            context.fill(handleX, sliderY - 2, handleX + 8, sliderY + 10, hovered ? 0xFFFFFFFF : 0xFFC8C8C8);

            // Value display
            String valueText = String.format("%d", (int)(value * 255));
            int valueX = sliderX + SLIDER_WIDTH + 10;
            context.drawTextWithShadow(textRenderer, valueText, valueX, labelY, 0xFFFFFF);
        }

        private int getPreviewColor() {
            int val = (int)(value * 255);
            if (channelName.equals("Red")) {
                return 0xFF000000 | (val << 16);
            } else if (channelName.equals("Green")) {
                return 0xFF000000 | (val << 8);
            } else {
                return 0xFF000000 | val;
            }
        }

        @Override
        public boolean click(int mouseX, int mouseY, int x, int y) {
            int sliderX = x + 10;
            int sliderY = y + 20;
            if (mouseX >= sliderX && mouseX <= sliderX + SLIDER_WIDTH &&
                mouseY >= sliderY - 2 && mouseY <= sliderY + 10) {
                updateValue(mouseX, sliderX);
                return true;
            }
            return false;
        }

        @Override
        public void drag(int mouseX, int x) {
            int sliderX = x + 10;
            updateValue(mouseX, sliderX);
        }

        private void updateValue(int mouseX, int sliderX) {
            int relativeX = Math.max(0, Math.min(SLIDER_WIDTH, mouseX - sliderX));
            value = relativeX / (float) SLIDER_WIDTH;
            onChange.run();
        }

        @Override
        public int getHeight() { return CONTROL_HEIGHT; }
    }

    private static class PreviewControl implements Control {
        final String label;

        PreviewControl(String label) {
            this.label = label;
        }

        @Override
        public void render(DrawContext context, int x, int y, boolean hovered, int mouseX) {
            int bgColor = 0xCC242424;
            context.fill(x, y, x + CONTROL_WIDTH, y + 45, bgColor);
            context.drawBorder(x, y, CONTROL_WIDTH, 45, 0xFF404040);

            net.minecraft.client.font.TextRenderer textRenderer = net.minecraft.client.MinecraftClient.getInstance().textRenderer;
            int labelX = x + 10;
            int labelY = y + 5;
            context.drawTextWithShadow(textRenderer, label, labelX, labelY, 0xFFFFFF);

            // Preview text
            String previewText = NameHider.getReplacementName();
            if (previewText.isEmpty()) previewText = "Preview";

            int previewY = y + 22;
            if (NameHider.isUsingGradient()) {
                // Draw gradient preview
                drawGradientText(context, textRenderer, previewText, x + CONTROL_WIDTH / 2, previewY);
            } else {
                // Draw solid color preview
                int color = rgbToInt(NameHider.getRed1(), NameHider.getGreen1(), NameHider.getBlue1());
                int textWidth = textRenderer.getWidth(previewText);
                context.drawTextWithShadow(textRenderer, previewText, x + (CONTROL_WIDTH - textWidth) / 2, previewY, 0xFF000000 | color);
            }
        }

        private void drawGradientText(DrawContext context, net.minecraft.client.font.TextRenderer textRenderer, String text, int centerX, int y) {
            int totalWidth = textRenderer.getWidth(text);
            int currentX = centerX - totalWidth / 2;

            for (int i = 0; i < text.length(); i++) {
                float ratio = text.length() == 1 ? 0.5f : (float) i / (text.length() - 1);
                float r = NameHider.getRed1() + (NameHider.getRed2() - NameHider.getRed1()) * ratio;
                float g = NameHider.getGreen1() + (NameHider.getGreen2() - NameHider.getGreen1()) * ratio;
                float b = NameHider.getBlue1() + (NameHider.getBlue2() - NameHider.getBlue1()) * ratio;

                int color = rgbToInt(r, g, b);
                String character = String.valueOf(text.charAt(i));
                context.drawTextWithShadow(textRenderer, character, currentX, y, 0xFF000000 | color);
                currentX += textRenderer.getWidth(character);
            }
        }

        private int rgbToInt(float r, float g, float b) {
            int ri = Math.max(0, Math.min(255, (int)(r * 255)));
            int gi = Math.max(0, Math.min(255, (int)(g * 255)));
            int bi = Math.max(0, Math.min(255, (int)(b * 255)));
            return (ri << 16) | (gi << 8) | bi;
        }

        @Override
        public boolean click(int mouseX, int mouseY, int x, int y) {
            return false;
        }

        @Override
        public void drag(int mouseX, int x) {}

        @Override
        public int getHeight() { return 45; }
    }

    public NameHiderCategoryScreen(Screen parent) {
        super(Text.literal("Name Hider Settings"));
        this.parent = parent;
    }

    @Override
    protected void init() {
        windowX = (this.width - WINDOW_WIDTH) / 2;
        windowY = (this.height - WINDOW_HEIGHT) / 2;

        if ((windowX & 1) != 0) windowX--;
        if ((windowY & 1) != 0) windowY--;

        // Create text field for replacement name
        nameField = new TextFieldWidget(this.textRenderer, 0, 0, CONTROL_WIDTH - 20, TEXT_FIELD_HEIGHT, Text.literal("Name"));
        nameField.setMaxLength(16);
        nameField.setText(NameHider.getReplacementName());
        nameField.setChangedListener(text -> {
            NameHider.setReplacementName(text);
        });
        this.addSelectableChild(nameField);

        setupControls();
    }

    private void setupControls() {
        controls = new ArrayList<>();

        controls.add(new ToggleControl("Enable Name Hider",
            NameHider.isEnabled(), () -> {
                NameHider.setEnabled(!NameHider.isEnabled());
                ((ToggleControl) controls.get(0)).enabled = NameHider.isEnabled();
            }));

        controls.add(new TextFieldControl("Replacement Name", nameField));

        controls.add(new ToggleControl("Use Gradient",
            NameHider.isUsingGradient(), () -> {
                NameHider.setUseGradient(!NameHider.isUsingGradient());
                ((ToggleControl) controls.get(2)).enabled = NameHider.isUsingGradient();
            }));

        controls.add(new ColorSliderControl("Color 1", "Red", NameHider.getRed1(), () -> {
            ColorSliderControl slider = (ColorSliderControl) controls.get(3);
            NameHider.setColor1(slider.value, NameHider.getGreen1(), NameHider.getBlue1());
        }));

        controls.add(new ColorSliderControl("Color 1", "Green", NameHider.getGreen1(), () -> {
            ColorSliderControl slider = (ColorSliderControl) controls.get(4);
            NameHider.setColor1(NameHider.getRed1(), slider.value, NameHider.getBlue1());
        }));

        controls.add(new ColorSliderControl("Color 1", "Blue", NameHider.getBlue1(), () -> {
            ColorSliderControl slider = (ColorSliderControl) controls.get(5);
            NameHider.setColor1(NameHider.getRed1(), NameHider.getGreen1(), slider.value);
        }));

        controls.add(new ColorSliderControl("Color 2", "Red", NameHider.getRed2(), () -> {
            ColorSliderControl slider = (ColorSliderControl) controls.get(6);
            NameHider.setColor2(slider.value, NameHider.getGreen2(), NameHider.getBlue2());
        }));

        controls.add(new ColorSliderControl("Color 2", "Green", NameHider.getGreen2(), () -> {
            ColorSliderControl slider = (ColorSliderControl) controls.get(7);
            NameHider.setColor2(NameHider.getRed2(), slider.value, NameHider.getBlue2());
        }));

        controls.add(new ColorSliderControl("Color 2", "Blue", NameHider.getBlue2(), () -> {
            ColorSliderControl slider = (ColorSliderControl) controls.get(8);
            NameHider.setColor2(NameHider.getRed2(), NameHider.getGreen2(), slider.value);
        }));

        controls.add(new PreviewControl("Preview"));
    }

    @Override
    public void renderBackground(DrawContext context, int mouseX, int mouseY, float delta) {
        this.renderInGameBackground(context);
    }

    @Override
    public void render(DrawContext context, int mouseX, int mouseY, float delta) {
        this.renderBackground(context, mouseX, mouseY, delta);

        // Main window background
        context.fill(windowX, windowY, windowX + WINDOW_WIDTH, windowY + WINDOW_HEIGHT, 0xCC000000);
        context.fill(windowX + 2, windowY + 2, windowX + WINDOW_WIDTH - 2, windowY + WINDOW_HEIGHT - 2, 0x88000000);

        // Borders
        context.drawBorder(windowX, windowY, WINDOW_WIDTH, WINDOW_HEIGHT, 0xFFFF9944);
        context.drawBorder(windowX + 1, windowY + 1, WINDOW_WIDTH - 2, WINDOW_HEIGHT - 2, 0x88FF9944);

        // Title
        int titleWidth = this.textRenderer.getWidth(this.title);
        int titleX = windowX + (WINDOW_WIDTH - titleWidth) / 2;
        int titleY = windowY + 15;
        context.drawTextWithShadow(this.textRenderer, this.title, titleX, titleY, 0xFFFFFF);

        drawControls(context, mouseX, mouseY);
        drawBackButton(context, mouseX, mouseY);

        super.render(context, mouseX, mouseY, delta);
    }

    private void drawControls(DrawContext context, int mouseX, int mouseY) {
        int startX = windowX + (WINDOW_WIDTH - CONTROL_WIDTH) / 2;
        int startY = windowY + 50;

        if ((startX & 1) != 0) startX--;
        if ((startY & 1) != 0) startY--;

        if (draggingSlider == -1) {
            hoveredControl = -1;
        }

        int currentY = startY;
        for (int i = 0; i < controls.size(); i++) {
            Control control = controls.get(i);
            int controlHeight = control.getHeight();

            boolean hovered = (draggingSlider == i) ||
                            (draggingSlider == -1 && mouseX >= startX && mouseX <= startX + CONTROL_WIDTH &&
                             mouseY >= currentY && mouseY <= currentY + controlHeight);

            if (hovered && draggingSlider == -1) hoveredControl = i;

            control.render(context, startX, currentY, hovered, mouseX);

            if (draggingSlider == i) {
                control.drag(mouseX, startX);
            }

            currentY += controlHeight + CONTROL_SPACING;
        }
    }

    private void drawBackButton(DrawContext context, int mouseX, int mouseY) {
        int buttonX = windowX + (WINDOW_WIDTH - BACK_BUTTON_WIDTH) / 2;
        int buttonY = windowY + WINDOW_HEIGHT - BACK_BUTTON_HEIGHT - 15;

        if ((buttonX & 1) != 0) buttonX--;
        if ((buttonY & 1) != 0) buttonY--;

        backButtonHovered = mouseX >= buttonX && mouseX <= buttonX + BACK_BUTTON_WIDTH &&
                           mouseY >= buttonY && mouseY <= buttonY + BACK_BUTTON_HEIGHT;

        int buttonColor = backButtonHovered ? 0xCC3A3A3A : 0xCC2A2A2A;
        context.fill(buttonX, buttonY, buttonX + BACK_BUTTON_WIDTH, buttonY + BACK_BUTTON_HEIGHT, buttonColor);
        context.drawBorder(buttonX, buttonY, BACK_BUTTON_WIDTH, BACK_BUTTON_HEIGHT,
                backButtonHovered ? 0xFFFF9944 : 0xFF404040);

        String buttonText = "Back";
        int textWidth = this.textRenderer.getWidth(buttonText);
        int textX = buttonX + (BACK_BUTTON_WIDTH - textWidth) / 2;
        int textY = buttonY + (BACK_BUTTON_HEIGHT - this.textRenderer.fontHeight) / 2;
        context.drawTextWithShadow(this.textRenderer, buttonText, textX, textY,
                backButtonHovered ? 0xFFFFFF : 0xC8C8C8);
    }

    @Override
    public boolean mouseClicked(double mouseX, double mouseY, int button) {
        if (button == 0) {
            if (backButtonHovered) {
                this.close();
                return true;
            }

            int startX = windowX + (WINDOW_WIDTH - CONTROL_WIDTH) / 2;
            int currentY = windowY + 50;

            boolean clickedAnyControl = false;
            for (int i = 0; i < controls.size(); i++) {
                Control control = controls.get(i);
                int controlHeight = control.getHeight();

                if (control.click((int) mouseX, (int) mouseY, startX, currentY)) {
                    if (control instanceof ColorSliderControl) {
                        draggingSlider = i;
                    }
                    clickedAnyControl = true;
                    // If we didn't click the text field, unfocus it
                    if (!(control instanceof TextFieldControl)) {
                        nameField.setFocused(false);
                    }
                    return true;
                }

                currentY += controlHeight + CONTROL_SPACING;
            }

            // If we clicked outside all controls, unfocus the text field
            if (!clickedAnyControl) {
                nameField.setFocused(false);
            }
        }
        return super.mouseClicked(mouseX, mouseY, button);
    }

    @Override
    public boolean mouseReleased(double mouseX, double mouseY, int button) {
        if (button == 0) {
            draggingSlider = -1;
        }
        return super.mouseReleased(mouseX, mouseY, button);
    }

    @Override
    public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
        if (nameField.keyPressed(keyCode, scanCode, modifiers)) {
            return true;
        }
        return super.keyPressed(keyCode, scanCode, modifiers);
    }

    @Override
    public boolean charTyped(char chr, int modifiers) {
        if (nameField.charTyped(chr, modifiers)) {
            return true;
        }
        return super.charTyped(chr, modifiers);
    }

    @Override
    public void close() {
        if (MiningqolClient.getConfig() != null) {
            MiningqolClient.getConfig().loadFromGame();
            MiningqolClient.getConfig().save();
        }

        if (this.client != null) {
            this.client.setScreen(parent);
        }
    }

    @Override
    public boolean shouldPause() {
        return false;
    }
}

File: src/client/kotlin/forfun/miningqol/client/gui/NameHiderCategoryScreen.kt
package forfun.miningqol.client.gui

import forfun.miningqol.client.MiningqolClient
import forfun.miningqol.client.NameHider
import net.minecraft.client.MinecraftClient
import net.minecraft.client.gui.screen.Screen
import net.minecraft.client.gui.widget.TextFieldWidget
import net.minecraft.text.Text
import xyz.meowing.knit.api.input.KnitKeys
import xyz.meowing.vexel.core.VexelScreen
import xyz.meowing.vexel.components.core.Rectangle
import xyz.meowing.vexel.components.core.Text as VText
import xyz.meowing.vexel.elements.Button
import xyz.meowing.vexel.components.base.Pos
import xyz.meowing.vexel.components.base.Size
import xyz.meowing.vexel.animations.*

class NameHiderCategoryScreen(private val parentScreen: Screen) : VexelScreen("Name Hider Settings") {
    private lateinit var overlay: Rectangle
    private lateinit var mainPanel: Rectangle
    private lateinit var nameField: TextFieldWidget
    private val mc = MinecraftClient.getInstance()

    override fun afterInitialization() {
        // Initialize text field
        nameField = TextFieldWidget(mc.textRenderer, 0, 0, 460, 20, Text.literal("Name"))
        nameField.setMaxLength(16)
        nameField.text = NameHider.getReplacementName()
        nameField.setChangedListener { text ->
            NameHider.setReplacementName(text)
        }

        // Semi-transparent dark overlay background
        overlay = Rectangle(
            backgroundColor = 0x80000000.toInt(),
            borderColor = 0x00000000,
            borderRadius = 0f,
            borderThickness = 0f
        )
            .setSizing(100f, Size.ParentPerc, 100f, Size.ParentPerc)
            .setPositioning(0f, Pos.ParentPixels, 0f, Pos.ParentPixels)
            .childOf(window)
            .fadeIn(400, EasingType.EASE_OUT)

        // Main panel - darker and more modern
        mainPanel = Rectangle(
            backgroundColor = 0xF0121212.toInt(),
            borderColor = 0xFF2A2A2A.toInt(),
            borderRadius = 16f,
            borderThickness = 1f
        )
            .setSizing(550f, Size.Pixels, 550f, Size.Pixels)
            .childOf(window)
            .apply {
                dropShadow = true
                shadowBlur = 40f
                shadowSpread = 2f
                shadowColor = 0xA0000000.toInt()
            }

        // Center the panel manually
        mainPanel.xPositionConstraint = Pos.ScreenPixels
        mainPanel.yPositionConstraint = Pos.ScreenPixels
        mainPanel.xConstraint = (mainPanel.screenWidth - 550f) / 2f
        mainPanel.yConstraint = (mainPanel.screenHeight - 550f) / 2f
        mainPanel.fadeIn(500, EasingType.EASE_OUT)

        // Title bar background
        Rectangle(
            backgroundColor = 0xFF1A1A1A.toInt(),
            borderColor = 0xFF2A2A2A.toInt(),
            borderRadius = 16f,
            borderThickness = 0f
        )
            .setSizing(100f, Size.ParentPerc, 80f, Size.Pixels)
            .setPositioning(0f, Pos.ParentPixels, 0f, Pos.ParentPixels)
            .childOf(mainPanel)
            .apply {
                borderRadiusBottomLeft = 0f
                borderRadiusBottomRight = 0f
            }
            .fadeIn(600, EasingType.EASE_OUT)

        // Title
        VText("Name Hider Settings", 0xFFFFFFFF.toInt(), 28f, true)
            .setPositioning(0f, Pos.ParentCenter, 18f, Pos.ParentPixels)
            .childOf(mainPanel)
            .fadeIn(700, EasingType.EASE_OUT)

        // Subtitle
        VText("Customize player name replacement and colors", 0xFF888888.toInt(), 13f, false)
            .setPositioning(0f, Pos.ParentCenter, 50f, Pos.ParentPixels)
            .childOf(mainPanel)
            .fadeIn(800, EasingType.EASE_OUT)

        // Enable toggle
        createToggleCard(
            "Enable Name Hider",
            0xFFFF9944.toInt(),
            { NameHider.isEnabled() },
            { NameHider.setEnabled(!NameHider.isEnabled()) },
            35f,
            100f,
            480f,
            65f,
            mainPanel,
            200L
        )

        // Replacement name text field
        createTextFieldCard(
            "Replacement Name",
            35f,
            180f,
            480f,
            65f,
            mainPanel,
            300L
        )

        // Gradient toggle
        createToggleCard(
            "Use Gradient",
            0xFFFF9944.toInt(),
            { NameHider.isUsingGradient() },
            { NameHider.setUseGradient(!NameHider.isUsingGradient()) },
            35f,
            260f,
            480f,
            65f,
            mainPanel,
            400L
        )

        // Preview card
        createPreviewCard(
            35f,
            340f,
            480f,
            80f,
            mainPanel,
            500L
        )

        // Note about color settings
        VText("Color sliders not yet implemented - use Java screen for now", 0xFF666666.toInt(), 11f, false)
            .setPositioning(0f, Pos.ParentCenter, 0f, Pos.ParentPixels)
            .alignBottom()
            .setOffset(0f, -75f)
            .childOf(mainPanel)
            .fadeIn(900, EasingType.EASE_OUT)

        // Back button at bottom
        Button("Back", 0xFFFFFFFF.toInt(), fontSize = 15f)
            .setSizing(140f, Size.Pixels, 42f, Size.Pixels)
            .setPositioning(0f, Pos.ParentCenter, 0f, Pos.ParentPixels)
            .alignBottom()
            .setOffset(0f, -25f)
            .backgroundColor(0xFF2A2A2A.toInt())
            .borderColor(0xFF404040.toInt())
            .borderRadius(8f)
            .borderThickness(1f)
            .hoverColors(0xFF353535.toInt(), 0xFFFFFFFF.toInt())
            .pressedColors(0xFF1A1A1A.toInt(), 0xFFAAAAAA.toInt())
            .onClick { _, _, _ ->
                closeWithAnimation()
                true
            }
            .childOf(mainPanel)
            .fadeIn(1000, EasingType.EASE_OUT)
    }

    private fun createToggleCard(
        label: String,
        accentColor: Int,
        getEnabled: () -> Boolean,
        toggleAction: () -> Unit,
        x: Float,
        y: Float,
        width: Float,
        height: Float,
        parent: Rectangle,
        animDelay: Long
    ) {
        val card = Rectangle(
            backgroundColor = 0xF01E1E1E.toInt(),
            borderColor = 0xFF2A2A2A.toInt(),
            borderRadius = 12f,
            borderThickness = 1f,
            hoverColor = 0xF0252525.toInt()
        )
            .setSizing(width, Size.Pixels, height, Size.Pixels)
            .setPositioning(x, Pos.ParentPixels, y, Pos.ParentPixels)
            .childOf(parent)
            .apply {
                dropShadow = true
                shadowBlur = 15f
                shadowSpread = 1f
                shadowColor = 0x40000000.toInt()
            }

        val enabled = getEnabled()
        val accentBar = Rectangle(
            backgroundColor = if (enabled) accentColor else 0xFF303030.toInt(),
            borderRadius = 12f
        )
            .setSizing(5f, Size.Pixels, 100f, Size.ParentPerc)
            .setPositioning(0f, Pos.ParentPixels, 0f, Pos.ParentPixels)
            .ignoreMouseEvents()
            .childOf(card)
            .apply {
                borderRadiusTopRight = 0f
                borderRadiusBottomRight = 0f
            }

        VText(label, 0xFFFFFFFF.toInt(), 20f, true)
            .setPositioning(20f, Pos.ParentPixels, 18f, Pos.ParentPixels)
            .childOf(card)

        val statusTextStr = if (enabled) "ON" else "OFF"
        val statusColor = if (enabled) accentColor else 0xFF606060.toInt()
        val statusText = VText(statusTextStr, statusColor, 16f, true)
            .setPositioning(20f, Pos.ParentPixels, 43f, Pos.ParentPixels)
            .childOf(card)

        card.onClick { _, _, _ ->
            toggleAction()
            val newEnabled = getEnabled()
            statusText.textColor = if (newEnabled) accentColor else 0xFF606060.toInt()
            statusText.text = if (newEnabled) "ON" else "OFF"
            accentBar.backgroundColor = if (newEnabled) accentColor else 0xFF303030.toInt()
            true
        }

        card.visible = false
        Thread {
            Thread.sleep(animDelay)
            mc.execute {
                card.fadeIn(400, EasingType.EASE_OUT)
            }
        }.start()
    }

    private fun createTextFieldCard(
        label: String,
        x: Float,
        y: Float,
        width: Float,
        height: Float,
        parent: Rectangle,
        animDelay: Long
    ) {
        val card = Rectangle(
            backgroundColor = 0xF01E1E1E.toInt(),
            borderColor = 0xFF2A2A2A.toInt(),
            borderRadius = 12f,
            borderThickness = 1f
        )
            .setSizing(width, Size.Pixels, height, Size.Pixels)
            .setPositioning(x, Pos.ParentPixels, y, Pos.ParentPixels)
            .childOf(parent)
            .apply {
                dropShadow = true
                shadowBlur = 15f
                shadowSpread = 1f
                shadowColor = 0x40000000.toInt()
            }

        VText(label, 0xFFFFFFFF.toInt(), 16f, true)
            .setPositioning(15f, Pos.ParentPixels, 12f, Pos.ParentPixels)
            .childOf(card)

        card.visible = false
        Thread {
            Thread.sleep(animDelay)
            mc.execute {
                card.fadeIn(400, EasingType.EASE_OUT)
            }
        }.start()
    }

    private fun createPreviewCard(
        x: Float,
        y: Float,
        width: Float,
        height: Float,
        parent: Rectangle,
        animDelay: Long
    ) {
        val card = Rectangle(
            backgroundColor = 0xF01E1E1E.toInt(),
            borderColor = 0xFF2A2A2A.toInt(),
            borderRadius = 12f,
            borderThickness = 1f
        )
            .setSizing(width, Size.Pixels, height, Size.Pixels)
            .setPositioning(x, Pos.ParentPixels, y, Pos.ParentPixels)
            .childOf(parent)
            .apply {
                dropShadow = true
                shadowBlur = 15f
                shadowSpread = 1f
                shadowColor = 0x40000000.toInt()
            }

        VText("Preview", 0xFFFFFFFF.toInt(), 16f, true)
            .setPositioning(15f, Pos.ParentPixels, 12f, Pos.ParentPixels)
            .childOf(card)

        // Note: Preview text rendering would need custom implementation
        VText("(Preview rendering not yet implemented)", 0xFF888888.toInt(), 12f, false)
            .setPositioning(0f, Pos.ParentCenter, 45f, Pos.ParentPixels)
            .childOf(card)

        card.visible = false
        Thread {
            Thread.sleep(animDelay)
            mc.execute {
                card.fadeIn(400, EasingType.EASE_OUT)
            }
        }.start()
    }

    override fun render(context: net.minecraft.client.gui.DrawContext, mouseX: Int, mouseY: Int, delta: Float) {
        super.render(context, mouseX, mouseY, delta)

        // Render text field on top
        val panelX = (width - 550f) / 2f
        val panelY = (height - 550f) / 2f
        nameField.setPosition((panelX + 50).toInt(), (panelY + 215).toInt())
        nameField.setWidth(460)
        nameField.render(context, mouseX, mouseY, delta)
    }

    override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
        nameField.mouseClicked(mouseX, mouseY, button)
        return super.mouseClicked(mouseX, mouseY, button)
    }

    override fun keyPressed(keyCode: Int, scanCode: Int, modifiers: Int): Boolean {
        if (nameField.keyPressed(keyCode, scanCode, modifiers)) {
            return true
        }
        return super.keyPressed(keyCode, scanCode, modifiers)
    }

    override fun charTyped(chr: Char, modifiers: Int): Boolean {
        if (nameField.charTyped(chr, modifiers)) {
            return true
        }
        return super.charTyped(chr, modifiers)
    }

    private fun closeWithAnimation() {
        MiningqolClient.getConfig()?.loadFromGame()
        MiningqolClient.getConfig()?.save()

        overlay.fadeOut(300, EasingType.EASE_IN)
        mainPanel.fadeOut(300, EasingType.EASE_IN) {
            mc.setScreen(parentScreen)
        }
    }

    override fun onKeyType(typedChar: Char, keyCode: Int, scanCode: Int) {
        if (keyCode == KnitKeys.KEY_ESCAPE.code) {
            closeWithAnimation()
        } else {
            super.onKeyType(typedChar, keyCode, scanCode)
        }
    }

    override fun shouldCloseOnEsc(): Boolean = false
}
